### PR TITLE
Donors/Samples/Datasets TSVs

### LINKED ISSUE #2039: Endpoints to grab Donor/Sample/Dataset metadata
Using the APIs is hard: people just want a TSV of the metadata. 

Proposal:
- `/api/v1/donors.tsv`, `/api/v1/samples.tsv`, `/api/v1/datasets.tsv`
- In addition to the "metadata" include IDs, creation date, creator.
- Order of columns will be arbitrary.
- Sorting will be arbitrary.
- column names can be the internal, ugly, ES, names.
- If logged in, more will be returned.
- Document this API.

@ngehlenborg : Look good? In particular:
- There are a lot of fields beyond the "metadata"... Do you just want IDs, creation date, creator... or would you like more? What about metadata from ancestor entities?
- If it's easier for people to ask us to re-implement the existing APIs, rather than using what's out there, then part of the problem may be documentation... So it follows that we should document our nascent API at least as well as the real APIs are documented.
- Bill should be told about this, at least so their API docs can be updated to point at the easy-to-use APIs we're implementing.

### PR DESCRIPTION
**Pull request by mccalluc**
Fix #2039: Metadata TSVs. Add an `/apis` link in the footer to a new page that describes the different APIs and data download options we have. (The url is not `/docs/apis` because we've entirely handed that directory off to PK, but I'd like to just get this described here.)
- @shirey : I would like your buy-in on this description of the existing APIs.
- @ngehlenborg : Confirmation that this is what you want.
- @john-conroy : Code review.

Current headers of `donors.tsv`:
```
uuid
hubmap_id
age_unit
age_value
blood_type
body_mass_index_unit
body_mass_index_value
cause_of_death
death_event
height_unit
height_value
kidney_donor_profile_index_unit
kidney_donor_profile_index_value
mechanism_of_injury
medical_history
race
rh_factor
sex
social_history
weight_unit
weight_value
```

Current headers of `samples.tsv`:
```
uuid
hubmap_id
cold_ischemia_time_unit
cold_ischemia_time_value
health_status
organ_condition
pathologist_report
perfusion_solution
procedure_date
sample_id
specimen_preservation_temperature
specimen_quality_criteria
specimen_tumor_distance_unit
specimen_tumor_distance_value
vital_state
warm_ischemia_time_unit
warm_ischemia_time_value
```

`datasets.tsv` has a huge number of columns... perhaps add an optional query parameter so users could just get a particular assay type? But I'd like to be more clear about the real user story, and I don't think that should block this PR.

### CODE REVIEW

File: context/app/api/client.py [hunk 1]
```diff
@@ -65,18 +62,34 @@ def get_all_dataset_uuids(self):
...
+            },
+            "_source": [*non_metadata_fields, 'mapped_metadata', 'metadata']
+        }
+        response_json = self._request(
+            current_app.config['ELASTICSEARCH_ENDPOINT']
+            + current_app.config['PORTAL_INDEX_PATH'],
+            body_json=query)
+        sources = [hit['_source'] for hit in response_json['hits']['hits']]
+        if len(sources) == size:
+            raise Exception('At least 10k datasets: need to make multiple requests')
```
Comment (john-conroy) [inline context/app/api/client.py:89]: If we're not enabling the means to change the result window couldn't we just use `hits.hits.total.value` and if it's greater than 10,000 throw the exception?

File: context/app/markdown/apis.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# APIs and Data Downloads
+
+HuBMAP incorporates a wide range of data and metadata, developed by a number of separate teams,
+so unfortunately there is no single "HuBMAP API". Depending on your needs, there are different places to begin.
```
Comment (ngehlenborg) [inline context/app/markdown/apis.md:4]:
```suggestion
and there is no single "HuBMAP API". Depending on your needs, there are different places to begin.
```

File: context/app/markdown/apis.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+- Download the backing JSON by appending `.json` to the URL, or clicking the document icon in the upper right.
+- Download just the metadata as a key-value TSV by clicking the download button beside the "Metadata" section header.
+- Download the data files through Globus, via a link in the "Files" section.
+
+## Metadata for all donors, samples, or datasets
+
+For metadata in TSV form, visit [`/api/v0/donors.tsv`](/api/v0/donors.tsv),
+[`/api/v0/samples.tsv`](/api/v0/samples.tsv), or [`/api/v0/datasets.tsv`](/api/v0/datasets.tsv).
+The datasets TSV has a large number of columns, because different assay types have different metadata fields.
+If you are logged in and have sufficient privs, these will include unpublished data.
```
Comment (ngehlenborg) [inline context/app/markdown/apis.md:18]:
```suggestion
If you are logged in and have sufficient access privileges, these will include unpublished data.
```

File: context/app/markdown/apis.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+## Query metadata
+
+To more flexibly query metadata, an [Elasticsearch endpoint](https://smart-api.info/ui/7aaf02b838022d564da776b03f357158) is available.
+- If you haven't used it before, Elasticsearch query syntax is non-trivial.
```
Comment (ngehlenborg) [inline context/app/markdown/apis.md:23]:
```suggestion
- Elasticsearch query syntax is non-trivial and only recommended for expert users.
```

File: context/app/markdown/apis.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+- The response structure is not documented, and we do not guarantee the stability of responses.
+
+## Create datasets
+
+[All API microservices](/docs/apis) developed by the IEC are documented at
+[SmartAPI](https://smart-api.info/registry?q=hubmap).
+These APIs support our internal UIs, and are unlikely to be useful to most external users.
```
Comment (ngehlenborg) [inline context/app/markdown/apis.md:30]:
```suggestion
These APIs support our internal UIs, and are likely of limited value to most external users.
```

File: context/app/markdown/apis.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+## Query gene expression data
+
+It is possible to query gene expression data across datasets.
+The [API](https://github.com/hubmapconsortium/cross_modality_query#usage) is complex and not well documented,
```
Comment (ngehlenborg) [inline context/app/markdown/apis.md:35]:
```suggestion
The [API](https://github.com/hubmapconsortium/cross_modality_query#usage) is complex,
```

**Review by ngehlenborg — CHANGES_REQUESTED** [review]
Thanks! Just some minor edits.

**Comment by mccalluc** [comment]
@ngehlenborg -- @shirey believes strongly that this should be part of their API, and they'll be able to do this in the next sprint. I'm not interested in having a tug-of-war about it. Closing.

**Comment by mccalluc** [comment]
@shirey has said he will suggest more changes, and then approve.

File: context/app/markdown/apis.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+## Metadata for all donors, samples, or datasets
+
+For metadata in TSV form, visit [`/api/v0/donors.tsv`](/api/v0/donors.tsv),
+[`/api/v0/samples.tsv`](/api/v0/samples.tsv), or [`/api/v0/datasets.tsv`](/api/v0/datasets.tsv).
+The datasets TSV has a large number of columns, because different assay types have different metadata fields.
+If you are logged in and have sufficient privs, these will include unpublished data.
+
+## Query metadata
+
+To more flexibly query metadata, an [Elasticsearch endpoint](https://smart-api.info/ui/7aaf02b838022d564da776b03f357158) is available.
```
Comment (shirey) [inline context/app/markdown/apis.md:22]:
```suggestion
To more flexibly query metadata, the [HuBMAP Search API](https://smart-api.info/ui/7aaf02b838022d564da776b03f357158) is available.
```

File: context/app/markdown/apis.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+- If you haven't used it before, Elasticsearch query syntax is non-trivial.
+- The response structure is not documented, and we do not guarantee the stability of responses.
+
+## Create datasets
```
Comment (shirey) [inline context/app/markdown/apis.md:26]:
```suggestion
## Edit HuBMAP provenance entities: datasets, tissue samples and donors
```

File: context/app/markdown/apis.md (hunk 5 above)
Comment (shirey) [inline context/app/markdown/apis.md]:
```suggestion
The  [HuBMAP Entity API](https://smart-api.info/ui/0065e419668f3336a40d1f5ab89c6ba3) can be used to create, update and retrieve provenance entities which consist of three base types donors, tissue samples and datasets.
Creating and updating of entities require internal access, but published entities can be programmatically retrieved by anyone. 
```

File: context/app/markdown/apis.md (hunk 6 above)
Comment (shirey) [inline context/app/markdown/apis.md]:
```suggestion
The [API](https://github.com/hubmapconsortium/cross_modality_query#usage) is complex,
```

**Review by shirey — CHANGES_REQUESTED** [review]

File: context/app/markdown/apis.md (hunk 7 above)
Comment (mccalluc) [reply context/app/markdown/apis.md:22]: Add missing period, back-ticks, https://www.w3.org/QA/Tips/noClickHere
```suggestion
To more flexibly query metadata, the [HuBMAP Search API](https://smart-api.info/ui/7aaf02b838022d564da776b03f357158) is available.
This web service is a thin wrapper around the [Elasticsearch `_search` method](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html).
```

File: context/app/markdown/apis.md (hunk 8 above)
Comment (mccalluc) [reply context/app/markdown/apis.md:26]:
```suggestion
## Edit HuBMAP datasets, samples, and donors
```

File: context/app/markdown/apis.md (hunk 5 above)
Comment (mccalluc) [reply context/app/markdown/apis.md]: Filed https://github.com/hubmapconsortium/portal-docs/issues/70: Should clean up docs if we don't think they're worth referencing.

**Pull request merged**